File: server/state.py
"""Learner model: single-writer store with a JSON snapshot and an append-only
JSONL event log. Every mastery change carries provenance (the event that
caused it). Only LearnerState mutates the model; everything else reads.

Mastery levels per concept: unseen -> exposed -> shaky -> mastered.
The known set (for fringe computation) = units whose terminal check passed the
gate OR were explicitly credited by a catch-up check.
"""

from __future__ import annotations

import json
import time
from pathlib import Path

LEVELS = ["unseen", "exposed", "shaky", "mastered"]


def _now() -> float:
    return time.time()


class LearnerState:
    def __init__(self, state_dir: str | Path, corpus):
        self.dir = Path(state_dir)
        self.dir.mkdir(parents=True, exist_ok=True)
        self.snapshot_path = self.dir / "learner.json"
        self.log_path = self.dir / "events.jsonl"
        self.corpus = corpus
        if self.snapshot_path.exists():
            self.data = json.loads(self.snapshot_path.read_text())
        else:
            self.data = self._fresh()
            self._save()

    def _fresh(self) -> dict:
        return {
            "version": 0,
            "concepts": {},          # id -> {level, history: [event refs], confidence: []}
            "units": {},             # id -> {status: locked|available|active|passed|overridden|debt,
                                     #        check_score, attempts}
            "misconceptions": {},    # id -> {evidence: [...], active: bool}
            "debt": [],              # [{unit, concepts, items_missed, reason, ts, retired}]
            "profile": {
                "expert_axes": ["code", "systems", "llm-tooling-behavior"],
                "novice_axes": ["matrix-calculus", "ml-notation"],
                "assumed_known": [],
            },
            "pacing": {"chunks": [], "breaks": [], "fatigue_flag": False,
                       "session_started": None},
            "summary": "New session. No evidence yet; treat per the static profile.",
            "current_unit": None,
            "chapter_cache": {},     # unit -> generated chapter payload (for resume)
        }

    # ---------- event-sourced mutation (the ONLY write path) ----------

    def apply(self, kind: str, payload: dict) -> dict:
        event = {"n": self.data["version"] + 1, "ts": _now(), "kind": kind, **payload}
        # The directory can disappear between boot and write (cleanup script,
        # redeploy, wiped volume). Losing a learner's progress mid-flight to a
        # missing mkdir is not an acceptable failure, so re-create on demand.
        self.dir.mkdir(parents=True, exist_ok=True)
        with self.log_path.open("a") as f:
            f.write(json.dumps(event) + "\n")
        self.data["version"] = event["n"]
        handler = getattr(self, f"_on_{kind}", None)
        if handler:
            handler(event)
        self._save()
        return event

    def _save(self):
        self.dir.mkdir(parents=True, exist_ok=True)
        self.snapshot_path.write_text(json.dumps(self.data, indent=1))

    # ---------- event handlers ----------

    def _concept(self, cid: str) -> dict:
        return self.data["concepts"].setdefault(
            cid, {"level": "unseen", "history": [], "confidence": []})

    def _set_level(self, cid: str, level: str, event_n: int, why: str):
        c = self._concept(cid)
        if c["level"] != level:
            c["history"].append({"n": event_n, "from": c["level"], "to": level, "why": why})
            c["level"] = level

    def _on_exposed(self, ev):
        for cid in ev["concepts"]:
            if self._concept(cid)["level"] == "unseen":
                self._set_level(cid, "exposed", ev["n"], f"read chapter {ev['unit']}")

    def _on_item_graded(self, ev):
        cid, verdict, conf = ev["concept"], ev["verdict"], ev.get("confidence")
        c = self._concept(cid)
        if conf is not None:
            c["confidence"].append({"n": ev["n"], "confidence": conf,
                                    "correct": verdict in ("pass", "valid_alternative_path")})
        ok = verdict in ("pass", "valid_alternative_path")
        cur = c["level"]
        if ok and cur in ("unseen", "exposed", "shaky"):
            # one correct answer never jumps straight to mastered from cold
            nxt = "shaky" if cur in ("unseen", "exposed") else "mastered"
            self._set_level(cid, nxt, ev["n"], f"correct on {ev['item']}")
        elif not ok and cur == "mastered":
            self._set_level(cid, "shaky", ev["n"], f"missed {ev['item']}")
        elif not ok and cur in ("unseen", "exposed"):
            self._set_level(cid, "shaky", ev["n"], f"missed {ev['item']} (needs work)")
        for mid in ev.get("misconceptions", []):
            m = self.data["misconceptions"].setdefault(mid, {"evidence": [], "active": True})
            m["active"] = True
            m["evidence"].append({"n": ev["n"], "item": ev["item"],
                                  "quote": ev.get("evidence", "")})

    def _on_misconception_cleared(self, ev):
        m = self.data["misconceptions"].get(ev["id"])
        if m:
            m["active"] = False
            m["evidence"].append({"n": ev["n"], "cleared_by": ev.get("why", "")})

    def _on_check_result(self, ev):
        u = self.data["units"].setdefault(ev["unit"], {"attempts": 0})
        u["attempts"] = u.get("attempts", 0) + 1
        u["check_score"] = ev["score"]
        u["status"] = "passed" if ev["passed"] else "failed"
        if ev["passed"]:
            for cid in ev.get("mastered_concepts", []):
                self._set_level(cid, "mastered", ev["n"], f"passed {ev['unit']} check @ {ev['score']:.0%}")

    def _on_override(self, ev):
        self.data["units"].setdefault(ev["unit"], {})["status"] = "overridden"
        self.data["debt"].append({
            "unit": ev["unit"], "concepts": ev["concepts"],
            "items_missed": ev.get("items_missed", []),
            "reason": ev["reason"],  # failed_gate | skipped_check
            "ts": _now(), "retired": False,
        })

    def _on_debt_retired(self, ev):
        for d in self.data["debt"]:
            if d["unit"] == ev["unit"] and not d["retired"]:
                d["retired"] = True

    def _on_chunk(self, ev):
        self.data["pacing"]["chunks"].append(
            {"n": ev["n"], "minutes": ev["minutes"], "unit": ev.get("unit")})

    def _on_break_taken(self, ev):
        self.data["pacing"]["breaks"].append({"n": ev["n"], "minutes": ev["minutes"]})
        self.data["pacing"]["fatigue_flag"] = False

    def _on_fatigue(self, ev):
        self.data["pacing"]["fatigue_flag"] = True

    def _on_summary(self, ev):
        self.data["summary"] = ev["text"]

    def _on_unit_started(self, ev):
        self.data["current_unit"] = ev["unit"]
        self.data["units"].setdefault(ev["unit"], {})["status"] = "active"
        if self.data["pacing"]["session_started"] is None:
            self.data["pacing"]["session_started"] = _now()

    def _on_chapter_cached(self, ev):
        self.data["chapter_cache"][ev["unit"]] = ev["ref"]

    def _on_assumed_known(self, ev):
        known = set(self.data["profile"]["assumed_known"])
        known.update(ev["concepts"])
        self.data["profile"]["assumed_known"] = sorted(known)

    # ---------- read-side queries ----------

    def unit_status(self, unit_id: str) -> str:
        return self.data["units"].get(unit_id, {}).get("status", "locked")

    def cleared_units(self) -> set:
        return {u for u, d in self.data["units"].items()
                if d.get("status") in ("passed", "overridden")}

    def fringe(self) -> list:
        """ALEKS-style outer fringe: units whose prereqs are all cleared and
        which aren't cleared themselves. Overridden units count as cleared for
        availability (agency) - the debt ledger carries the difference."""
        graph = self.corpus.prereq_graph()
        cleared = self.cleared_units()
        loaded = set(self.corpus.units)
        out = []
        for uid in self.corpus.unit_order():
            if uid in cleared or uid not in loaded:
                continue  # never offer a unit whose files are missing
            if all(p in cleared for p in graph.get(uid, []) if p in loaded):
                out.append(uid)
        return out

    def active_misconceptions(self) -> list:
        return [mid for mid, m in self.data["misconceptions"].items() if m["active"]]

    def open_debt(self) -> list:
        return [d for d in self.data["debt"] if not d["retired"]]

    def concept_level(self, cid: str) -> str:
        return self.data["concepts"].get(cid, {}).get("level", "unseen")

    def fragile_concepts(self) -> list:
        """Inner fringe: recently mastered or shaky - the review queue for
        cumulative-check callbacks. Low-confidence-correct also lands here."""
        out = []
        for cid, c in self.data["concepts"].items():
            if c["level"] == "shaky":
                out.append(cid)
            elif c["level"] == "mastered" and c["confidence"]:
                last = c["confidence"][-1]
                if last["correct"] and last["confidence"] <= 2:
                    out.append(cid)
        return out

    def session_minutes(self) -> float:
        start = self.data["pacing"]["session_started"]
        return (time.time() - start) / 60 if start else 0.0

    def minutes_since_break(self) -> float:
        breaks = self.data["pacing"]["breaks"]
        chunks = self.data["pacing"]["chunks"]
        if not chunks:
            return 0.0
        return sum(c["minutes"] for c in chunks[-4:]) if not breaks else \
            sum(c["minutes"] for c in chunks if c["n"] > breaks[-1]["n"])

File: server-go/state/state.go
// Package state holds the learner model: a single-writer store with a JSON
// snapshot and an append-only JSONL event log. Every mastery change carries
// provenance - the event that caused it. Only this package mutates the model;
// everything else reads.
//
// Mastery per concept: unseen -> exposed -> shaky -> mastered. The known set
// (for fringe computation) is units whose terminal check passed the gate, or
// which were explicitly credited by a catch-up check.
//
// The snapshot format is byte-compatible with the Python implementation's, so
// an existing learner.json keeps working across the rewrite.
package state

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"sync"
	"time"

	"github.com/mjbraun/chiron/server/corpus"
)

type Concept struct {
	Level      string            `json:"level"`
	History    []LevelChange     `json:"history"`
	Confidence []ConfidencePoint `json:"confidence"`
}

type LevelChange struct {
	N    int    `json:"n"`
	From string `json:"from"`
	To   string `json:"to"`
	Why  string `json:"why"`
}

type ConfidencePoint struct {
	N          int  `json:"n"`
	Confidence int  `json:"confidence"`
	Correct    bool `json:"correct"`
}

type UnitState struct {
	Status     string   `json:"status,omitempty"`
	CheckScore *float64 `json:"check_score,omitempty"`
	Attempts   int      `json:"attempts"`
}

type MisconceptionState struct {
	Evidence []map[string]any `json:"evidence"`
	Active   bool             `json:"active"`
}

type Debt struct {
	Unit        string   `json:"unit"`
	Concepts    []string `json:"concepts"`
	ItemsMissed []string `json:"items_missed"`
	Reason      string   `json:"reason"` // failed_gate | skipped_check
	TS          float64  `json:"ts"`
	Retired     bool     `json:"retired"`
}

type Profile struct {
	ExpertAxes   []string `json:"expert_axes"`
	NoviceAxes   []string `json:"novice_axes"`
	AssumedKnown []string `json:"assumed_known"`
	// SelfRating is the learner's own placement from the calibration
	// screener: 1 (absolute novice) to 5 (expert). 0 = not asked yet.
	SelfRating int `json:"self_rating,omitempty"`
}

type Chunk struct {
	N       int     `json:"n"`
	Minutes float64 `json:"minutes"`
	Unit    string  `json:"unit,omitempty"`
}

type Break struct {
	N       int     `json:"n"`
	Minutes float64 `json:"minutes"`
}

type Pacing struct {
	Chunks         []Chunk  `json:"chunks"`
	Breaks         []Break  `json:"breaks"`
	FatigueFlag    bool     `json:"fatigue_flag"`
	SessionStarted *float64 `json:"session_started"`
}

type Data struct {
	Version        int                            `json:"version"`
	Concepts       map[string]*Concept            `json:"concepts"`
	Units          map[string]*UnitState          `json:"units"`
	Misconceptions map[string]*MisconceptionState `json:"misconceptions"`
	Debt           []*Debt                        `json:"debt"`
	Profile        Profile                        `json:"profile"`
	Pacing         Pacing                         `json:"pacing"`
	Summary        string                         `json:"summary"`
	CurrentUnit    *string                        `json:"current_unit"`
	ChapterCache   map[string]any                 `json:"chapter_cache"`
}

// Event is one entry in the append-only log. Fields are optional per kind; the
// zero values are never written thanks to omitempty, so the log stays readable.
type Event struct {
	N    int     `json:"n"`
	TS   float64 `json:"ts"`
	Kind string  `json:"kind"`

	Unit             string   `json:"unit,omitempty"`
	Concepts         []string `json:"concepts,omitempty"`
	Item             string   `json:"item,omitempty"`
	Concept          string   `json:"concept,omitempty"`
	Verdict          string   `json:"verdict,omitempty"`
	Confidence       *int     `json:"confidence,omitempty"`
	Misconceptions   []string `json:"misconceptions,omitempty"`
	Evidence         string   `json:"evidence,omitempty"`
	ID               string   `json:"id,omitempty"`
	Why              string   `json:"why,omitempty"`
	Score            *float64 `json:"score,omitempty"`
	Passed           *bool    `json:"passed,omitempty"`
	MasteredConcepts []string `json:"mastered_concepts,omitempty"`
	ItemsMissed      []string `json:"items_missed,omitempty"`
	Reason           string   `json:"reason,omitempty"`
	Minutes          *float64 `json:"minutes,omitempty"`
	// Band is the calibration band (1-5) of a graded item; zero outside
	// calibration units.
	Band int    `json:"band,omitempty"`
	Text string `json:"text,omitempty"`
	Ref  any    `json:"ref,omitempty"`
}

type Learner struct {
	dir    string
	corpus *corpus.Corpus
	mu     sync.Mutex
	Data   *Data
	nowFn  func() float64
}

func now() float64 { return float64(time.Now().UnixNano()) / 1e9 }

func Open(dir string, c *corpus.Corpus) (*Learner, error) {
	l := &Learner{dir: dir, corpus: c, nowFn: now}
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return nil, err
	}
	raw, err := os.ReadFile(l.snapshotPath())
	switch {
	case err == nil:
		var d Data
		if err := json.Unmarshal(raw, &d); err != nil {
			return nil, fmt.Errorf("learner.json: %w", err)
		}
		l.Data = &d
		l.normalise()
	case os.IsNotExist(err):
		l.Data = fresh()
		if err := l.save(); err != nil {
			return nil, err
		}
	default:
		return nil, err
	}
	return l, nil
}

func (l *Learner) snapshotPath() string { return filepath.Join(l.dir, "learner.json") }
func (l *Learner) logPath() string      { return filepath.Join(l.dir, "events.jsonl") }

// LogPath exposes the append-only log so the review exporter can replay it.
func (l *Learner) LogPath() string { return l.logPath() }

func fresh() *Data {
	return &Data{
		Concepts:       map[string]*Concept{},
		Units:          map[string]*UnitState{},
		Misconceptions: map[string]*MisconceptionState{},
		Debt:           []*Debt{},
		Profile: Profile{
			ExpertAxes:   []string{"code", "systems", "llm-tooling-behavior"},
			NoviceAxes:   []string{"matrix-calculus", "ml-notation"},
			AssumedKnown: []string{},
		},
		Pacing:       Pacing{Chunks: []Chunk{}, Breaks: []Break{}},
		Summary:      "New session. No evidence yet; treat per the static profile.",
		ChapterCache: map[string]any{},
	}
}

// normalise fills in maps a hand-edited or older snapshot might be missing, so
// a nil map never panics on first write.
func (l *Learner) normalise() {
	d := l.Data
	if d.Concepts == nil {
		d.Concepts = map[string]*Concept{}
	}
	if d.Units == nil {
		d.Units = map[string]*UnitState{}
	}
	if d.Misconceptions == nil {
		d.Misconceptions = map[string]*MisconceptionState{}
	}
	if d.ChapterCache == nil {
		d.ChapterCache = map[string]any{}
	}
	if d.Debt == nil {
		d.Debt = []*Debt{}
	}
}

// ---------- event-sourced mutation (the ONLY write path) ----------

func (l *Learner) Apply(ev Event) (Event, error) {
	l.mu.Lock()
	defer l.mu.Unlock()

	ev.N = l.Data.Version + 1
	ev.TS = l.nowFn()

	// The directory can disappear between boot and write (cleanup script,
	// redeploy, wiped volume). Losing a learner's progress mid-flight to a
	// missing mkdir is not an acceptable failure, so re-create on demand.
	if err := os.MkdirAll(l.dir, 0o755); err != nil {
		return ev, err
	}
	line, err := json.Marshal(ev)
	if err != nil {
		return ev, err
	}
	f, err := os.OpenFile(l.logPath(), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return ev, err
	}
	if _, err := f.Write(append(line, '\n')); err != nil {
		f.Close()
		return ev, err
	}
	if err := f.Close(); err != nil {
		return ev, err
	}

	l.Data.Version = ev.N
	l.handle(ev)
	return ev, l.save()
}

func (l *Learner) save() error {
	if err := os.MkdirAll(l.dir, 0o755); err != nil {
		return err
	}
	out, err := json.MarshalIndent(l.Data, "", " ")
	if err != nil {
		return err
	}
	// Write-then-rename: a crash mid-write must not leave a truncated snapshot
	// where a learner's whole history used to be.
	tmp := l.snapshotPath() + ".tmp"
	if err := os.WriteFile(tmp, out, 0o644); err != nil {
		return err
	}
	return os.Rename(tmp, l.snapshotPath())
}

// ---------- event handlers ----------

func (l *Learner) handle(ev Event) {
	switch ev.Kind {
	case "exposed":
		for _, cid := range ev.Concepts {
			if l.concept(cid).Level == "unseen" {
				l.setLevel(cid, "exposed", ev.N, "read chapter "+ev.Unit)
			}
		}
	case "item_graded":
		l.onItemGraded(ev)
	case "misconception_cleared":
		if m, ok := l.Data.Misconceptions[ev.ID]; ok {
			m.Active = false
			m.Evidence = append(m.Evidence, map[string]any{"n": ev.N, "cleared_by": ev.Why})
		}
	case "check_result":
		l.onCheckResult(ev)
	case "override":
		u := l.unit(ev.Unit)
		u.Status = "overridden"
		l.Data.Debt = append(l.Data.Debt, &Debt{
			Unit: ev.Unit, Concepts: ev.Concepts, ItemsMissed: ev.ItemsMissed,
			Reason: ev.Reason, TS: l.nowFn(), Retired: false,
		})
	case "debt_retired":
		for _, d := range l.Data.Debt {
			if d.Unit == ev.Unit && !d.Retired {
				d.Retired = true
			}
		}
	case "chunk":
		m := 0.0
		if ev.Minutes != nil {
			m = *ev.Minutes
		}
		l.Data.Pacing.Chunks = append(l.Data.Pacing.Chunks, Chunk{N: ev.N, Minutes: m, Unit: ev.Unit})
	case "break_taken":
		m := 0.0
		if ev.Minutes != nil {
			m = *ev.Minutes
		}
		l.Data.Pacing.Breaks = append(l.Data.Pacing.Breaks, Break{N: ev.N, Minutes: m})
		l.Data.Pacing.FatigueFlag = false
	case "fatigue":
		l.Data.Pacing.FatigueFlag = true
	case "summary":
		l.Data.Summary = ev.Text
	case "unit_started":
		unit := ev.Unit
		l.Data.CurrentUnit = &unit
		l.unit(ev.Unit).Status = "active"
		if l.Data.Pacing.SessionStarted == nil {
			t := l.nowFn()
			l.Data.Pacing.SessionStarted = &t
		}
	case "chapter_cached":
		l.Data.ChapterCache[ev.Unit] = ev.Ref
	case "self_rating":
		if ev.Score != nil {
			l.Data.Profile.SelfRating = int(*ev.Score)
		}
	case "assumed_known":
		seen := map[string]bool{}
		for _, c := range l.Data.Profile.AssumedKnown {
			seen[c] = true
		}
		for _, c := range ev.Concepts {
			seen[c] = true
		}
		out := make([]string, 0, len(seen))
		for c := range seen {
			out = append(out, c)
		}
		sort.Strings(out)
		l.Data.Profile.AssumedKnown = out
	}
}

func (l *Learner) onItemGraded(ev Event) {
	c := l.concept(ev.Concept)
	ok := ev.Verdict == "pass" || ev.Verdict == "valid_alternative_path"
	if ev.Confidence != nil {
		c.Confidence = append(c.Confidence, ConfidencePoint{
			N: ev.N, Confidence: *ev.Confidence, Correct: ok,
		})
	}
	switch cur := c.Level; {
	case ok && (cur == "unseen" || cur == "exposed"):
		// One correct answer never jumps straight to mastered from cold.
		l.setLevel(ev.Concept, "shaky", ev.N, "correct on "+ev.Item)
	case ok && cur == "shaky" && ev.Band > 0 && ev.Band < 3:
		// A floor probe - arithmetic, recognition, a spelled-out recipe -
		// is evidence the learner is not lost, never that the concept is
		// held. Mastery waits for an item at the level of the concept itself.
	case ok && cur == "shaky":
		l.setLevel(ev.Concept, "mastered", ev.N, "correct on "+ev.Item)
	case !ok && cur == "mastered":
		l.setLevel(ev.Concept, "shaky", ev.N, "missed "+ev.Item)
	case !ok && (cur == "unseen" || cur == "exposed"):
		l.setLevel(ev.Concept, "shaky", ev.N, "missed "+ev.Item+" (needs work)")
	}
	for _, mid := range ev.Misconceptions {
		m, ok := l.Data.Misconceptions[mid]
		if !ok {
			m = &MisconceptionState{Evidence: []map[string]any{}}
			l.Data.Misconceptions[mid] = m
		}
		m.Active = true
		m.Evidence = append(m.Evidence, map[string]any{
			"n": ev.N, "item": ev.Item, "quote": ev.Evidence,
		})
	}
}

func (l *Learner) onCheckResult(ev Event) {
	u := l.unit(ev.Unit)
	u.Attempts++
	if ev.Score != nil {
		s := *ev.Score
		u.CheckScore = &s
	}
	passed := ev.Passed != nil && *ev.Passed
	if passed {
		u.Status = "passed"
		score := 0.0
		if ev.Score != nil {
			score = *ev.Score
		}
		why := fmt.Sprintf("passed %s check @ %.0f%%", ev.Unit, score*100)
		for _, cid := range ev.MasteredConcepts {
			l.setLevel(cid, "mastered", ev.N, why)
		}
	} else {
		u.Status = "failed"
	}
}

func (l *Learner) concept(cid string) *Concept {
	c, ok := l.Data.Concepts[cid]
	if !ok {
		c = &Concept{Level: "unseen", History: []LevelChange{}, Confidence: []ConfidencePoint{}}
		l.Data.Concepts[cid] = c
	}
	return c
}

func (l *Learner) unit(uid string) *UnitState {
	u, ok := l.Data.Units[uid]
	if !ok {
		u = &UnitState{}
		l.Data.Units[uid] = u
	}
	return u
}

func (l *Learner) setLevel(cid, level string, eventN int, why string) {
	c := l.concept(cid)
	if c.Level == level {
		return
	}
	c.History = append(c.History, LevelChange{N: eventN, From: c.Level, To: level, Why: why})
	c.Level = level
}

// Reset clears the learner model back to a fresh session.
//
// The event log is truncated rather than appended to: it is the provenance
// record for the state that exists, and keeping a previous run's grades under a
// snapshot that no longer reflects them makes the review export lie about what
// was missed. The old log is kept beside it, once, so a reset by mistake is
// recoverable.
func (l *Learner) Reset() error {
	l.mu.Lock()
	defer l.mu.Unlock()

	if _, err := os.Stat(l.logPath()); err == nil {
		if err := os.Rename(l.logPath(), l.logPath()+".previous"); err != nil {
			return err
		}
	}
	l.Data = fresh()
	return l.save()
}

// ---------- read-side queries ----------

func (l *Learner) UnitStatus(unitID string) string {
	l.mu.Lock()
	defer l.mu.Unlock()
	if u, ok := l.Data.Units[unitID]; ok && u.Status != "" {
		return u.Status
	}
	return "locked"
}

func (l *Learner) ClearedUnits() map[string]bool {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.clearedLocked()
}

func (l *Learner) clearedLocked() map[string]bool {
	out := map[string]bool{}
	for id, u := range l.Data.Units {
		if u.Status == "passed" || u.Status == "overridden" {
			out[id] = true
		}
	}
	return out
}

// Fringe is the ALEKS-style outer fringe: units whose prerequisites are all
// cleared and which are not cleared themselves. An overridden unit counts as
// cleared for availability - that is the learner's agency, and the debt ledger
// carries the difference.
func (l *Learner) Fringe() []string {
	l.mu.Lock()
	defer l.mu.Unlock()
	graph := l.corpus.PrereqGraph()
	cleared := l.clearedLocked()
	var out []string
	for _, uid := range l.corpus.UnitOrder() {
		if cleared[uid] {
			continue
		}
		if _, loaded := l.corpus.Units[uid]; !loaded {
			continue // never offer a unit whose files are missing
		}
		ready := true
		for _, p := range graph[uid] {
			if _, loaded := l.corpus.Units[p]; !loaded {
				continue
			}
			if !cleared[p] {
				ready = false
				break
			}
		}
		if ready {
			out = append(out, uid)
		}
	}
	return out
}

func (l *Learner) ActiveMisconceptions() []string {
	l.mu.Lock()
	defer l.mu.Unlock()
	var out []string
	for mid, m := range l.Data.Misconceptions {
		if m.Active {
			out = append(out, mid)
		}
	}
	sort.Strings(out) // map order is random in Go; prompts must be stable
	return out
}

func (l *Learner) OpenDebt() []*Debt {
	l.mu.Lock()
	defer l.mu.Unlock()
	var out []*Debt
	for _, d := range l.Data.Debt {
		if !d.Retired {
			out = append(out, d)
		}
	}
	return out
}

func (l *Learner) ConceptLevel(cid string) string {
	l.mu.Lock()
	defer l.mu.Unlock()
	if c, ok := l.Data.Concepts[cid]; ok {
		return c.Level
	}
	return "unseen"
}

// FragileConcepts is the inner fringe: shaky, or mastered but answered
// correctly with low confidence. This is the review queue that cumulative
// callbacks draw from - a low-confidence-correct answer is the fluency illusion
// showing, and LLM prose amplifies it.
func (l *Learner) FragileConcepts() []string {
	l.mu.Lock()
	defer l.mu.Unlock()
	var out []string
	for cid, c := range l.Data.Concepts {
		switch {
		case c.Level == "shaky":
			out = append(out, cid)
		case c.Level == "mastered" && len(c.Confidence) > 0:
			last := c.Confidence[len(c.Confidence)-1]
			if last.Correct && last.Confidence <= 2 {
				out = append(out, cid)
			}
		}
	}
	sort.Strings(out)
	return out
}

func (l *Learner) SessionMinutes() float64 {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.Data.Pacing.SessionStarted == nil {
		return 0
	}
	return (l.nowFn() - *l.Data.Pacing.SessionStarted) / 60
}

func (l *Learner) MinutesSinceBreak() float64 {
	l.mu.Lock()
	defer l.mu.Unlock()
	chunks, breaks := l.Data.Pacing.Chunks, l.Data.Pacing.Breaks
	if len(chunks) == 0 {
		return 0
	}
	total := 0.0
	if len(breaks) == 0 {
		start := max(0, len(chunks)-4)
		for _, c := range chunks[start:] {
			total += c.Minutes
		}
		return total
	}
	lastBreak := breaks[len(breaks)-1].N
	for _, c := range chunks {
		if c.N > lastBreak {
			total += c.Minutes
		}
	}
	return total
}

func (l *Learner) Snapshot() Data {
	l.mu.Lock()
	defer l.mu.Unlock()
	return *l.Data
}
